File: Python/Medium/1415 The k-th Lexicographical String of All Happy Strings of Length n.py
class Solution:
    def getHappyString(self, n: int, k: int) -> str:
        self.k = k
        self.result = ""
        
        def backtrack(current_path):
            if len(current_path) == n:
                self.k -= 1
                if self.k == 0:
                    self.result = "".join(current_path)
                return
            
            for char in ['a', 'b', 'c']:
                if self.result: return # Found the k-th string
                
                if not current_path or current_path[-1] != char:
                    current_path.append(char)
                    backtrack(current_path)
                    current_path.pop()
        
        backtrack([])
        return self.result

File: Java/Medium/1415 The k-th Lexicographical String of All Happy Strings of Length n.java
class Solution {
    int count = 0;
    String result = "";

    public String getHappyString(int n, int k) {
        count = 0;
        result = "";
        backtrack(n, k, new StringBuilder());
        return result;
    }

    private void backtrack(int n, int k, StringBuilder current) {
        if (current.length() == n) {
            count++;
            if (count == k) {
                result = current.toString();
            }
            return;
        }

        for (char ch : new char[]{'a', 'b', 'c'}) {
            if (result.length() > 0) return; // Optimization: stop if found
            
            if (current.length() == 0 || current.charAt(current.length() - 1) != ch) {
                current.append(ch);
                backtrack(n, k, current);
                current.deleteCharAt(current.length() - 1);
            }
        }
    }
}
